sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseIO.java

```diff
@@ -0,0 +1,364 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Properties;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.io.clickhouse.TableSchema.ColumnType;
+import org.apache.beam.sdk.io.clickhouse.TableSchema.DefaultType;
+import org.apache.beam.sdk.schemas.FieldAccessDescriptor;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PDone;
+import org.apache.beam.sdk.values.Row;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import ru.yandex.clickhouse.ClickHouseConnection;
+import ru.yandex.clickhouse.ClickHouseDataSource;
+import ru.yandex.clickhouse.ClickHouseStatement;
+import ru.yandex.clickhouse.settings.ClickHouseQueryParam;
+
+/** An IO to write to ClickHouse. */
+@Experimental(Experimental.Kind.SCHEMAS)
+public class ClickHouseIO {
+
+  /** A {@link PTransform} to write to ClickHouse. */
+  @AutoValue
+  public abstract static class Write<T> extends PTransform<PCollection<T>, PDone> {
+    public abstract String jdbcUrl();
+
+    public abstract String table();
+
+    @Nullable
+    public abstract Properties properties();
+
+    public static TableSchema getTableSchema(String jdbcUrl, String table) {
+      ResultSet rs;
+      try (ClickHouseConnection connection = new ClickHouseDataSource(jdbcUrl).getConnection();
+          Statement statement = connection.createStatement()) {
+        rs = statement.executeQuery("DESCRIBE TABLE " + table);
+        List<TableSchema.Column> columns = new ArrayList<>();
+
+        while (rs.next()) {
+          String name = rs.getString("name");
+          String type = rs.getString("type");
+          String defaultTypeStr = rs.getString("default_type");
+          String defaultExpression = rs.getString("default_expression");
+
+          ColumnType columnType = ColumnType.parse(type);
+          DefaultType defaultType = DefaultType.parse(defaultTypeStr).orElse(null);
+
+          Object defaultValue;
+          if (DefaultType.DEFAULT.equals(defaultType)
+              && !Strings.isNullOrEmpty(defaultExpression)) {
+            defaultValue = ColumnType.parseDefaultExpression(columnType, defaultExpression);
+          } else {
+            defaultValue = null;
+          }
+
+          columns.add(TableSchema.Column.of(name, columnType, defaultType, defaultValue));
+        }
+
+        // findbugs doesn't like it in try block
+        rs.close();
```

Comment:
I think this should be in finally block.

Reply:
Initially I did inner try-with-resources, but it doesn't work with findbugs for some reason.